File: 76.最小覆盖子串.py
#
# @lc app=leetcode.cn id=76 lang=python
#
# [76] 最小覆盖子串
#

# TLE
# @lc code=start
class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        if len(s) < len(t):
            return ""

        def contains(s1, s2):
            m = {}
            for i in s2:
                if i in m:
                    m[i] += 1
                else:
                    m[i] = 1
            for i in s1:
                if i in m:
                    m[i] -= 1
            for v in m.values():
                if v > 0:
                    return False
            return True

        l1, l2, res = 0, len(t), ""
        while l2 <= len(s):
            if l2 - l1 >= len(t):
                if contains(s[l1:l2], t):
                    if len(s[l1:l2]) == len(t):
                        return s[l1:l2]
                    if len(s[l1:l2]) < len(res) or res == "":
                        res = s[l1:l2]
                        l1 += 1
                    else:
                        l1 += 1
                else:
                    l2 += 1
        return res

# @lc code=end

File: 76.最小覆盖子串.go
/*
 * @lc app=leetcode.cn id=76 lang=golang
 *
 * [76] 最小覆盖子串
 */

// @lc code=start
func containChars(a, b string) bool {
	mapb := map[rune]int{}
	for _, val := range(b) {
			if _, ok := mapb[val]; ok {
					mapb[val]++
			} else {
					mapb[val] = 1
			}
	}
	
	for _, val := range(a) {
			if _, ok := mapb[val]; ok {
					if mapb[val] > 0 {
							mapb[val]--
					}
			}
	}
	
	for _, val := range(mapb) {
			if val > 0 {
					return false
			}   
	}
	return true
}

func minWindow(s string, t string) string {
	len1, len2 := len(s), len(t)
	if len2 > len1 {
			return ""
	}
	l, r, res := 0, len2, ""
	for {
			if r <= len1 {
					var target string
					if (r - l) >= len2 {
							target = s[l:r]
							if containChars(target, t) {
									if len(target) == len2 {
											return target
									}
									if len(target) < len(res) || res == "" {
											res = target
									} else {
											l++
									}
							} else {
									r++
							}
					} else {
							break
					}
			} else {
					break
			}
	}
	return res
}
// @lc code=end

